From 925bb8a7b7daf67e6d7d1a7dba48752bd360e498 Mon Sep 17 00:00:00 2001
From: Ikey Doherty <ikey.doherty@gmail.com>
Date: Thu, 10 Jul 2014 19:44:38 +0100
Subject: [PATCH 13/22] wm: Link "budgie-panel --menu" to panel-main-menu
 keybinding

On Evolve OS this will result in the main menu being invoked when someone
uses the <Super_L> key, it is controlled by the GNOME keybindings prefs.
---
 wm/plugin.c | 16 ++++++++++++++++
 wm/plugin.h |  2 ++
 2 files changed, 18 insertions(+)

diff --git a/wm/plugin.c b/wm/plugin.c
index 7b3dd20..e8d0572 100644
--- a/wm/plugin.c
+++ b/wm/plugin.c
@@ -39,6 +39,7 @@
 #include <meta/meta-background-group.h>
 #include <meta/meta-background-actor.h>
 #include <meta/prefs.h>
+#include <meta/keybindings.h>
 
 #include <libintl.h>
 #define _(x) dgettext (GETTEXT_PACKAGE, x)
@@ -150,6 +151,17 @@ typedef struct
   MetaPlugin *plugin;
 } EffectCompleteData;
 
+/* Budgie specific callbacks */
+void budgie_launch_menu (MetaDisplay    *display,
+                         MetaScreen     *screen,
+                         MetaWindow     *window,
+                         XIDeviceEvent  *event,
+                         MetaKeyBinding *binding,
+                         gpointer        user_data)
+{
+  /* Ask budgie-panel to open the menu */
+  g_spawn_command_line_async("budgie-panel --menu", NULL);
+}
 
 static void
 meta_default_plugin_dispose (GObject *object)
@@ -326,6 +338,10 @@ show_stage (MetaPlugin *plugin)
 
   clutter_actor_show (stage);
 
+  /* Set up our own keybinding overrides */
+  meta_keybindings_set_custom_handler(BUDGIE_KEYBINDING_MAIN_MENU,
+                                      budgie_launch_menu, NULL, NULL);
+
   return FALSE;
 }
 
diff --git a/wm/plugin.h b/wm/plugin.h
index 438f5b1..b9ded37 100644
--- a/wm/plugin.h
+++ b/wm/plugin.h
@@ -48,6 +48,8 @@ typedef struct _MetaDefaultPluginPrivate MetaDefaultPluginPrivate;
 #define MUTTER_EDGE_TILING "edge-tiling"
 #define MUTTER_MODAL_ATTACH "attach-modal-dialogs"
 
+#define BUDGIE_KEYBINDING_MAIN_MENU "panel-main-menu"
+
 struct _MetaDefaultPlugin
 {
   MetaPlugin parent;
-- 
1.9.1

